src/lib.rs

```diff
@@ -0,0 +1,186 @@
+}
+
+impl Tracer {
+    /// Create a new tracer.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// Tracer.new().trace_filename("mytrace.ptt").data_bufsize(1024).target_pid(666);
+    /// ```
+    pub fn new() -> Self {
+        Tracer {
+            trace_filename: CString::new("traceme.ptt").unwrap(),
+            target_pid: unsafe { getpid() },
+            tracer_ctx: None,
+            data_bufsize: 64,
+            aux_bufsize: 1024,
+        }
+    }
+
+    /// Set where to write the trace.
+    ///
+    /// # Arguments
+    ///
+    /// * `filename` - Filename to use.
+    pub fn trace_filename(mut self, filename: &str) -> Self {
+        self.trace_filename = CString::new(filename).expect("bad trace filename");
+        self
+    }
+
+    /// Select which PID to trace.
+    ///
+    /// By default, the current PID is traced.
+    ///
+    /// # Arguments
+    ///
+    /// * `pid` - PID to trace.
+    pub fn target_pid(mut self, pid: pid_t) -> Self {
+        self.target_pid = pid;
+        self
+    }
+
+    /// Set the PT data buffer size.
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - The size to use.
+    pub fn data_bufsize(mut self, size: usize) -> Self {
+        self.data_bufsize = size;
+        self
+    }
+
+    /// Set the PT aux buffer size.
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - The size to use.
+    pub fn aux_bufsize(mut self, size: usize) -> Self {
+        self.aux_bufsize = size;
+        self
+    }
+
+    // Make the map filename by setting/adding a ".map" extension to the trace filename
+    fn make_map_filename(&self, trace_filename: &CString) -> CString {
+        let map_string = trace_filename.clone().into_string().unwrap();
+        let mut map_path = PathBuf::from(map_string);
+        if !map_path.set_extension("map") {
+            panic!("failed to construct map filename");
+        }
+        let map_string = map_path.to_str().and_then(|ms| CString::new(ms).ok());
+        map_string.expect("failed to construct map filename")
+    }
+
+    /// Records execution of the selected PID into the chosen output file.
+    ///
+    /// Tracing continues until [stop_tracing](struct.Tracer.html#method.stop_tracing) is called.
+    pub fn start_tracing(&mut self) {
+        if self.tracer_ctx.is_some() {
+            panic!("This tracer has already been started");
+        }
+
+        // Build the C configuration struct
+        let map_filename_c = self.make_map_filename(&self.trace_filename);
+        let tr_conf = TracerConf {
+            target_pid: self.target_pid,
+            trace_filename: self.trace_filename.as_ptr(),
+            map_filename: map_filename_c.as_ptr(),
+            data_bufsize: self.data_bufsize,
+            aux_bufsize: self.aux_bufsize,
+        };
+
+        // Call C
+        let conf_ptr = &tr_conf as *const TracerConf;
+        let opq_ptr = unsafe {
+            // By transmuting we avoid the need to make TracerConf public.
```

Comment:
Isn't there some way that doesn't involve transmuting?

Reply:
Heh, I knew you would flag this.

I spent a fair amount of time investigating this on Friday. It used to be the case that passing a pointer out to C would mean the type being pointed to would need to be public. This no longer seems to be the case after running rustup again, so I'll push a fix.